src/target/riscv/riscv.c

```diff
@@ -2278,6 +2279,67 @@ derive_debug_reason_without_hitbit(const struct target *target, riscv_reg_t dpc)
 	}
 	return DBG_REASON_WATCHPOINT;
 }
+
+static bool check_single_step_needed_in_hit_supported_case(struct target *target,
+			uint32_t unique_id)
+{
+	RISCV_INFO(r);
+	riscv_reg_t tselect;
+	if (riscv_reg_get(target, &tselect, GDB_REGNO_TSELECT) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_reg_set(target, GDB_REGNO_TSELECT, unique_id) != ERROR_OK)
+		return ERROR_FAIL;
+	riscv_reg_t tdata1;
+	if (riscv_reg_get(target, &tdata1, GDB_REGNO_TDATA1) != ERROR_OK)
+		return ERROR_FAIL;
+	int type = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+	if (type == CSR_TDATA1_TYPE_MCONTROL) {
+		if (get_field(tdata1, CSR_MCONTROL_TIMING) == CSR_MCONTROL_TIMING_BEFORE)
+			r->need_single_step = true;
+	} else if (type == CSR_TDATA1_TYPE_MCONTROL6) {
+		int hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+		int hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+		int trigger_retired_info = (hit1 << 1) | hit0;
+		if (trigger_retired_info == CSR_MCONTROL6_HIT0_BEFORE)
+			r->need_single_step = true;
+	} else {
+		r->need_single_step = false;
+	}
+	if (riscv_reg_set(target, GDB_REGNO_TSELECT, tselect) != ERROR_OK)
+		return ERROR_FAIL;
+	return r->need_single_step;
+}
+
+static bool check_single_step_needed_in_hit_unsupported_case(struct target *target)
```

Comment:
Can this be avoided?
I'd suggest just conservatively setting `r->need_single_step = true;` if `hit` fields are not supported.
This will also probably be faster -- one step operation vs a bunch of register reads via JTAG.

Reply:
avoided, and addressed